enterprise/cluster/src/main/java/org/neo4j/cluster/logging/AsyncLogging.java

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (c) 2002-2015 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.cluster.logging;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadFactory;
+
+import org.neo4j.concurrent.AsyncEventSender;
+import org.neo4j.concurrent.AsyncEvents;
+import org.neo4j.function.Consumer;
+import org.neo4j.kernel.lifecycle.LifeSupport;
+import org.neo4j.kernel.lifecycle.LifecycleAdapter;
+import org.neo4j.logging.Log;
+import org.neo4j.logging.LogProvider;
+import org.neo4j.logging.NullLogProvider;
+import org.neo4j.logging.async.AsyncLogEvent;
+import org.neo4j.logging.async.AsyncLogProvider;
+
+import static org.neo4j.logging.async.AsyncLogEvent.process;
+import static org.neo4j.unsafe.impl.internal.dragons.FeatureToggles.flag;
+
+public class AsyncLogging extends LifecycleAdapter
+        implements Consumer<AsyncLogEvent>, ThreadFactory, AsyncEvents.Monitor
```

Comment:
I would extract Executors/ThreadFactory with `RerunOnFailure` into separate class to split up pure logging and `nonfailurable` executor